pkg/koordlet/metricsadvisor/collector_gpu.go

```diff
@@ -0,0 +1,191 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metricsadvisor
+
+import (
+	"sort"
+
+	"github.com/NVIDIA/go-nvml/pkg/nvml"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/metriccache"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/klog/v2"
+)
+
+type gpuUsageDetailRecord struct {
+	deviceCount      int
+	devices          []*device
+	processesMetrics map[uint32][]*rawGPUMetric
+}
+
+type rawGPUMetric struct {
+	SMUtil     uint32 // current utilization rate for the device
+	MemoryUsed uint64
+}
+
+type device struct {
+	Minor      int32 // index starting from 0
+	DeviceUUID string
+}
+
+func (g *gpuUsageDetailRecord) GetNodeGPUUsage() []*metriccache.GPUMetric {
+	tmp := make([]*rawGPUMetric, g.deviceCount)
+	// init tmp array.
+	for i := 0; i < g.deviceCount; i++ {
+		tmp[i] = &rawGPUMetric{}
+	}
+	for _, p := range g.processesMetrics {
+		for idx := 0; idx < g.deviceCount; idx++ {
+			if m := p[uint32(idx)]; m != nil {
+				tmp[idx].SMUtil += p[uint32(idx)].SMUtil
+				tmp[idx].MemoryUsed += p[uint32(idx)].MemoryUsed
+			}
+		}
+	}
+	rtn := make([]*metriccache.GPUMetric, g.deviceCount)
+	for i := 0; i < g.deviceCount; i++ {
+		rtn[i] = &metriccache.GPUMetric{
+			DeviceUUID: g.devices[i].DeviceUUID,
+			Minor:      g.devices[i].Minor,
+			SMUtil:     tmp[i].SMUtil,
+			MemoryUsed: *resource.NewQuantity(int64(tmp[i].MemoryUsed), resource.BinarySI),
+		}
+	}
+	return rtn
+}
+
+func (g *gpuUsageDetailRecord) GetPIDsTotalGPUUsage(pids []uint64) []*metriccache.GPUMetric {
+	tmp := make(map[int]*rawGPUMetric)
+	for _, pid := range pids {
+		if metrics, exist := g.processesMetrics[uint32(pid)]; exist {
+			for idx, metric := range metrics {
+				if metric == nil {
+					continue
+				}
+				if _, found := tmp[idx]; !found {
+					tmp[idx] = &rawGPUMetric{}
+				}
+				tmp[idx].MemoryUsed += metric.MemoryUsed
+				tmp[idx].SMUtil += metric.SMUtil
+			}
+		}
+	}
+	if len(tmp) == 0 {
+		return nil
+	}
+	rtn := make([]*metriccache.GPUMetric, 0)
+	for i := 0; i < g.deviceCount; i++ {
+		if value, ok := tmp[i]; ok {
+			rtn = append(rtn, &metriccache.GPUMetric{
+				DeviceUUID: g.devices[i].DeviceUUID,
+				Minor:      g.devices[i].Minor,
+				SMUtil:     value.SMUtil,
+				MemoryUsed: *resource.NewQuantity(int64(value.MemoryUsed), resource.BinarySI),
+			})
+		}
+	}
+	return rtn
+}
+
+func (c *collector) collectGPUUsage() {
+	c.context.nodeGPUUsageDetail = nil
+	if ret := nvml.Init(); ret != nvml.SUCCESS {
```

Comment:
The initialization should get called just once? Could you please resolve it or add a FIXME comment?

Reply:
refactored this code.